log/log.go

```diff
@@ -0,0 +1,192 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"strings"
+)
+
+const (
+	// LevelLogError specified the log level as: ERROR
+	LevelLogError = 1
+
+	// LevelWarning specified the log level as: WARNING
+	LevelWarning = 2
+
+	// LevelLogInfo specified the log level as: INFO
+	LevelLogInfo = 3
+
+	// LevelLogDebug specified the log level as: DEBUG
+	LevelLogDebug = 4
+
+	// LevelLogVerbose specified the log level as: VERBOSE
+	// This is the same as Debug, but without the repeat filtering
+	LevelLogVerbose = 5
+
+	// Mixer is use as a cmdTag when mixer command is logging error
+	Mixer = "MIXER"
+	// Dnf is use as a cmdTag to when dnf command is logging error
+	Dnf = "DNF"
+	// Rpm is use as a cmdTag when rpm is logging error
+	Rpm = "RPM"
+	// Rpm2Archive is use as a cmdTag when rpm2archive is logging error
+	Rpm2Archive = "RPM2ARCHIVE"
+	// Tar is use as a cmdTag when tar is logging error
+	Tar = "TAR"
+	// BsDiff is use as a cmdTag when bsDiff is logging error
+	BsDiff = "BSDIFF"
+)
+
+var (
+	level      = LevelLogInfo
+	levelMap   = map[int]string{}
+	fileHandle *os.File
+	logging    = false
+	lineLast   string
+	lineCount  int
+	cmdMap     = map[string]bool{}
+)
+
+func init() {
+	levelMap[LevelLogError] = "LevelLogError"
+	levelMap[LevelWarning] = "LevelWarning"
+	levelMap[LevelLogInfo] = "LevelLogInfo"
+	levelMap[LevelLogDebug] = "LevelLogDebug"
+	levelMap[LevelLogVerbose] = "LevelLogVerbose"
+	cmdMap[Mixer] = true
+	cmdMap[Dnf] = true
+	cmdMap[Rpm] = true
+	cmdMap[Rpm2Archive] = true
+	cmdMap[Tar] = true
+	cmdMap[BsDiff] = true
+}
+
+// SetLogLevel sets the default log level to l
+func SetLogLevel(l int) {
+	if l < LevelLogError {
+		level = LevelLogError
+		logTag("WRN", Mixer, "Log Level '%d' too low, forcing to %s (%d)", l, levelMap[level], level)
+	} else if l > LevelLogVerbose {
+		level = LevelLogVerbose
+		logTag("WRN", Mixer, "Log Level '%d' too high, forcing to %s (%d)", l, levelMap[level], level)
+	} else {
+		level = l
+		Debug(Mixer, "Log Level set to %s (%d)", levelMap[level], l)
+	}
+}
+
+// SetOutputFilename ... sets the default log output to filename instead of stdout/stderr
+func SetOutputFilename(logFile string) (*os.File, error) {
+	var err error
+	fileHandle, err = os.OpenFile(logFile, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, err
+	}
+
+	log.SetOutput(fileHandle)
+	logging = true
+	return fileHandle, nil
+}
+
+// GetLogHandler ... returns the file handler of logging file
+func GetLogHandler() *os.File {
+	return fileHandle
+}
+
+func logTag(tag string, cmdTag, format string, a ...interface{}) {
+	// If there are no variable to pass to the format,
+	// then we can escape any % signs.
+	if len(a) < 1 {
+		format = strings.ReplaceAll(format, "%", "%%")
+	}
+
+	f := "[" + tag + "] " + "[" + cmdTag + "] " + format + "\n"
```

Comment:
I think the "\n" is creating blank lines in the log file. 
Try removing this "\n" and instead explicitly add them to the source strings that are passed to the log command. 
E.g. 
```
log.Info(log.Mixer, "Building bundles...\n")
log.Info(log.Mixer, "Installing all bundles to full chroot\n")
```

OR see if there is a way we can avoid having to pass "\n" to the source strings but also ensure that additional blank lines are not being printed in the log file.